src/chains/ethereum/ethereum/src/helpers/simulations.ts

```diff
@@ -0,0 +1,583 @@
+  // the "best". so we do this here rather than in the common setup functions
+  if (common.isActivatedEIP(2930) && transaction.accessList) {
+    warmAccessList(stateManager, transaction.accessList);
+  }
+
+  const callResult = await _runCall({
+    blockchain,
+    runCallOpts,
+    vm,
+    options,
+    emitStepEvent
+  });
+
+  const callResultValue = callResult.execResult.returnValue;
+
+  return callResultValue === undefined
+    ? Data.Empty
+    : Data.from(callResultValue);
+};
+
+/**
+ * Clones the VM and iteratively simulates the transaction, generating an access
+ * list after for each run. The function returns once the generated access list
+ * is the same as that of the previous run or as the `initialAccessList`.
+ * @param initialAccessList An access list to compare against the generated
+ * access list.
+ * @returns The final access list generated, plus an estimate of the gas
+ * consumed by running the transaction _with_ the generated access list included.
+ */
+export const createAccessList = async ({
+  blockchain,
+  common,
+  simulationBlock,
+  transaction,
+  options,
+  emitStepEvent
+}: CreateAccessListOptions): Promise<CreateAccessListResult> => {
+  const { vm, stateManager, stateTrie } = await cloneVm({
+    blockchain,
+    common,
+    simulationBlock
+  });
+
+  const {
+    runCallOpts,
+    intrinsicGas,
+    accessListExclusions,
+    addressesOnlyStorage
+  } = await runCallSetup({ common, stateTrie, stateManager, transaction });
+
+  // no real reason why this is our max, feel free to change
+  const MAX_ITERATIONS = 1000;
+  let previousAccessList = transaction.accessList || [];
+  let iterations = 0;
+  do {
+    // checkpoint so we can get back to this vm state after every time we
+    // run this loop
+    await stateManager.checkpoint();
+    const callResult = await _runCall({
+      blockchain,
+      runCallOpts,
+      vm,
+      options,
+      emitStepEvent
+    });
+    const accessList = stateManager.generateAccessList(
+      accessListExclusions,
+      addressesOnlyStorage
+    );
+    // either the access list passed by the user is the same as what was
+    // generated by the VM, or the VM generated the same access list twice
+    // in a row, so this is our "best" access list.
+    if (AccessLists.areAccessListsSame(previousAccessList, accessList)) {
+      const { dataFeeEIP2930 } = AccessLists.getAccessListData(accessList);
+      const baseFeeBigInt = intrinsicGas + dataFeeEIP2930;
+      const gasUsedBigInt =
+        Quantity.toBigInt(callResult.gasUsed.toArrayLike(Buffer)) +
+        baseFeeBigInt;
+      const gasUsed = Quantity.from(gasUsedBigInt);
+      return { accessList, gasUsed };
+    } else {
+      // we are going to run this loop again so revert our changes
+      await stateManager.revert();
+      previousAccessList = accessList;
+    }
+    iterations++;
+  } while (iterations < MAX_ITERATIONS);
```

Comment:
Would it be worthwhile having a test case for this?

Also, in this case, shouldn't we also return the best-so-far access-list?

Reply:
Good find, we definitely should return if we hit max iterations. I've written a test for this, but it takes significant time to hit that many iterations, so it's not feasible for a test :/

Reply:
I figured that might be the case 👍